REGRESSION(264666@main): Layout of the <tspan> elements inside a <textPath> is incorrect

nikolaszimmermann opened the pull request:
#### 79238bc8e2d67087ed83ed550931b4c76964aa2a
<pre>
REGRESSION(264666@main): Layout of the &lt;tspan&gt; elements inside a &lt;textPath&gt; is incorrect
<a href="https://bugs.webkit.org/show_bug.cgi?id=264184">https://bugs.webkit.org/show_bug.cgi?id=264184</a>

Reviewed by Said Abou-Hallawa.

Revision 264666@main broke &lt;tspan textLength=&quot;...&quot;&gt; behavior on &lt;textPath&gt; elements.
The logic added to fix the intra-character spacing, wasn&apos;t intended for text path
layout, where each character by definition is placed in its own box. Fix the MDN
regression by disabling the new code paths for text-on-path layout.

Covered by new test in svg/text/textLength-tspan-in-textPath.html.

Unfortunately it&apos;s hard to make a cross-platform reftest, given the
nature of the kind of test, and the missing way to layout text along
a path in HTML/CSS.

* LayoutTests/platform/glib/svg/text/textLength-tspan-in-textPath-expected.txt: Added.
* LayoutTests/platform/ios/svg/text/textLength-tspan-in-textPath-expected.txt: Added.
* LayoutTests/platform/mac/svg/text/textLength-tspan-in-textPath-expected.png: Added.
* LayoutTests/platform/mac/svg/text/textLength-tspan-in-textPath-expected.txt: Added.
* LayoutTests/svg/text/textLength-tspan-in-textPath.html: Added.
* Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp:
(WebCore::SVGTextLayoutEngine::layoutTextOnLineOrPath):

Canonical link: <a href="https://commits.webkit.org/271996@main">https://commits.webkit.org/271996@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2e09cb49249deda19b1691b1ee73098b3a5582b7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29757 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8417 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/31059 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32259 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26904 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/30379 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10604 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5684 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/26926 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/30036 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/7051 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/25371 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5993 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/6172 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/26458 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33595 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/27147 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26882 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32343 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6082 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4280 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30125 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7828 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/26257 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7172 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6834 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6613 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp:
```diff
@@ -469,6 +469,9 @@ void SVGTextLayoutEngine::layoutTextOnLineOrPath(SVGInlineTextBox& textBox, Rend
             auto currentBoxOnLineHasAbsolutePosition = [&]() {
+                // In path layout, each character is absolutely positioned.
+                if (m_inPathLayout)
+                    return true;
```
I think this decision is deeply hidden and hard to see. Disabling the new code for the text-on-path-layout will be inside a lambda which is inside 3 nested if-statements. Can't we move this if-statement earlier:

```
if (m_visualCharacterOffset == textBox.start()) {
    if (m_inPathLayout)
        startsNewTextChunk = logicalAttributes->context().characterStartsNewTextChunk(m_logicalCharacterOffset);
    else {
        // The new code or a call to a new function like
        startsNewTextChunk = characterStartsNewTextChunkOnLine(logicalAttributes);
    }
    textBox.setStartsNewTextChunk(startsNewTextChunk);
}
```

nikolaszimmermann replied:
Your concerns are valid, there should be en easier-to-grasp solution. Will check.

nikolaszimmermann replied:
@shallawa pushed a new version, that strives to be easier to understand.

shallawa commented on Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp:
```diff
@@ -455,47 +455,60 @@ void SVGTextLayoutEngine::layoutTextOnLineOrPath(SVGInlineTextBox& textBox, Rend
+        // textLength to determine the chunk start position, instead of glyph advance values.
+        auto moveToExpectedChunkStartPositionIfNeeded = [&]() {
+            if (m_inPathLayout || !m_lastChunkHasTextLength || !previousBoxOnLine)
+                return;
```
Is it okay to bail out even the text position is equal to SVGTextLayoutAttributes::emptyValue() in these case?

shallawa commented on Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp:
```diff
@@ -455,47 +455,60 @@ void SVGTextLayoutEngine::layoutTextOnLineOrPath(SVGInlineTextBox& textBox, Rend
+            }
+
+            auto* textContentElement = SVGTextContentElement::elementFromRenderer(&previousBoxOnLine->renderer());
+            if (!textContentElement)
+                return;
```
When can this be true? Should an assertion fire here?

shallawa commented on Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp:
```diff
@@ -455,47 +455,60 @@ void SVGTextLayoutEngine::layoutTextOnLineOrPath(SVGInlineTextBox& textBox, Rend
-                    }
-                }
-            }
+            if (m_lastChunkHasTextLength && previousBoxOnLine)
+                return true;
```
Maybe this check can be moved before the check above since this one is faster and  `characterStartsNewTextChunk()`  does not have any side effect.

shallawa reviewed (approved):
Looks good to me. The review comments can be addressed later.

nikolaszimmermann committed:
REGRESSION(264666@main): Layout of the <tspan> elements inside a <textPath> is incorrect
https://bugs.webkit.org/show_bug.cgi?id=264184

Reviewed by Said Abou-Hallawa.

Revision 264666@main broke <tspan textLength="..."> behavior on <textPath> elements.
The logic added to fix the intra-character spacing, wasn't intended for text path
layout, where each character by definition is placed in its own box. Fix the MDN
regression by disabling the new code paths for text-on-path layout.

Covered by new test in svg/text/textLength-tspan-in-textPath.html.

Unfortunately it's hard to make a cross-platform reftest, given the
nature of the kind of test, and the missing way to layout text along
a path in HTML/CSS.

* LayoutTests/platform/glib/svg/text/textLength-tspan-in-textPath-expected.txt: Added.
* LayoutTests/platform/ios/svg/text/textLength-tspan-in-textPath-expected.txt: Added.
* LayoutTests/platform/mac/svg/text/textLength-tspan-in-textPath-expected.png: Added.
* LayoutTests/platform/mac/svg/text/textLength-tspan-in-textPath-expected.txt: Added.
* LayoutTests/svg/text/textLength-tspan-in-textPath.html: Added.
* Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp:
(WebCore::SVGTextLayoutEngine::layoutTextOnLineOrPath):

Canonical link: https://commits.webkit.org/271996@main
